src/PostgREST/AppState.hs

```diff
@@ -91,17 +94,58 @@ initWithPool newPool conf =
     <*> newIORef 0
 
 destroy :: AppState -> IO ()
-destroy = releasePool
+destroy = destroyPool
 
 initPool :: AppConfig -> IO SQL.Pool
 initPool AppConfig{..} =
-  SQL.acquire (configDbPoolSize, configDbPoolTimeout, toUtf8 configDbUri)
-
+  SQL.acquire configDbPoolSize $ toUtf8 configDbUri
+
+-- | Run an action with a database connection.
+--
+-- This is a thin wrapper around 'SQL.use', which deals with a
+-- race on flushing the pool: The pool read via 'readIORef'
+-- might be released via 'flushPool' by the time we call 'SQL.use',
+-- so we catch that 'SQL.PoolIsReleasedUsageError' and retry.
+--
+-- If on the other hand the pool has been (permanently) destroyed,
+-- we re-use 'SQL.PoolIsReleasedUsageError' to convey this to the
+-- caller.
 usePool :: AppState -> SQL.Session a -> IO (Either SQL.UsageError a)
-usePool AppState{..} = SQL.use statePool
-
-releasePool :: AppState -> IO ()
-releasePool AppState{..} = SQL.release statePool
+usePool appState@AppState{..} session = do
+  pool <- readIORef statePool
+  case pool of
+    Nothing -> return $ Left SQL.PoolIsReleasedUsageError
+    Just p -> do
+      res <- SQL.use p session
+      case res of
+        Left SQL.PoolIsReleasedUsageError -> do
+          logWithZTime appState "Attempted to used flushed pool, retrying"
+          usePool appState session
+        _                                 -> return res
+
+-- | Flush the connection pool so that any future use of the pool will
+-- use connections freshly established after this call. In-use
+-- connections aren't affected, they just won't be reused anymore.
```

Comment:
There's one issue with the way we replace pools, in that there's a timespan where two pools are around (the old one with the in-process requests, and the new empty one). Until all the old requests are done, it's possible to exceed the configured pool size.

Reply:
I've implemented this as a patch to `hasql-pool` here https://github.com/nikita-volkov/hasql-pool/pull/16.

Reply:
Just to be clear, does https://github.com/nikita-volkov/hasql-pool/pull/16 prevent exceeding the configured pool size?

Reply:
Yes.

> Just to be clear, does [nikita-volkov/hasql-pool#16](https://github.com/nikita-volkov/hasql-pool/pull/16) prevent exceeding the configured pool size?

Yes (well unless there's a bug in the PR).